src/momento/internal/aio/_scs_grpc_manager.py

```diff
@@ -64,8 +67,27 @@ def __init__(self, configuration: Configuration, credential_provider: Credential
                 # (experimental.ChannelOptions.SingleThreadedUnaryStream, 1)
             ],
         )
+        # we submit the eager connection task in our current running event loop
+        try:
+            self._eager_connect_task = asyncio.get_running_loop().create_task(self._eagerly_connect())
```

Comment:
This was a bit interesting. The `asyncio` semantics doesn't provide a subscribe method similar to the sync gRPC channel. This part is non-blocking, and required some hacks (without changing the constructor to be async) to make it blocking per the timeout. So in essence this is best effort. Lmk if you think we should make the constructor async which will make it easier to block on this until a timeout.

Reply:
I don't know if you even can make the constructor async, I don't know if the language allows it.

This is why, in the JS SDK, @nand4011 made a separate async factory function for instantiating the client. And that is also why the eager connection timeout did not end up being a new config setting at all, but rather, a new argument to that factory function.

My sense is that we need to follow that same path for the Python async API at least, and then for symmetry we would probably want to also add a factory function for the sync client.

I think this feature is kind of a non-starter if there is no way for the caller to block until the connection is established.

@malandis - thoughts?

Reply:
> I don't know if you even can make the constructor async, I don't know if the language allows it.

I'll go over this once more but at a quick glance I saw some samples; might be interpreting it different though.

> I think this feature is kind of a non-starter if there is no way for the caller to block until the connection is established.

I honestly had the same instinct and hence mentioned best-effort but now since you also called it out, it's not that tricky to make it blocking with the factory pattern you mentioned.

> then for symmetry we would probably want to also add a factory function for the sync client.

For the sync client, the reason I didn't think of that pattern as de-facto was because in NodeJS we **had** to follow that pattern as we couldn't do it in the existing constructor. In DotNet, I believe we enable this feature for free. And I believe we wanted to do the same in the SDKs where it was doable, so that customers get eager connections for free.

Reply:
I actually think that if I could go back in time I might have done the .NET one differently. We're making a network call so it seems pretty reasonable for it to surface to the end user as an async thing that they get to decide how they want to handle.

But @malandis is our most experienced python dev so it would be good to get his input.

Reply:
Agreed. Since the init methods are not async, we should not be awaiting anything in them, even if in a roundabout way. A factory method is a better here.